webservice/s3_file_handler.py

```diff
@@ -0,0 +1,149 @@
+        :rtype: list
+        """
+        response = self.session.list_objects_v2(Bucket=bucket)
+        if response.has_key('Contents'):
+            L = [response['Contents'][x]['Key']
+                 for x in range(len(response['Contents']))]
+            return [str(r) for r in L]  # convert to utf-8 (is unicode)
+        else:
+            return []
+
+    def get_bucket_list(self):
+        """
+        :returns a list of all buckets in AWS account
+        :rtype: list"""
+        buckets = self.bucket.list_buckets()
+        return [bucket['Name'] for bucket in buckets['Buckets']]
+
+    def upload_object_to_bucket(self, bucket, filename, key):
+        """Uploads some object (e.g., a file) to an S3 bucket.
+        :param bucket: name of bucket in AWS account
+        :type bucket: str
+        :param filename: absolute name of the object to upload
+        :type filename: str
+        :param key: object name in bucket on S3
+        :type key: str
+        :returns: status code, and msg if not 200
+        :rtype: JSON
+        """
+
+        status_code = self.bucket_exists(bucket)
+        if status_code == 200:
+            try:
+                # If it throws no exception we assume all went well.
+                self.resource.meta.client.upload_file(
+                    filename, bucket, key)  # executes silently
+                return {'status_code': 200}
+            except ClientError as e:
+                return {
+                    'status_code':
+                    e.response['ResponseMetadata']['HTTPStatusCode']}
+        else:
+            return {'status_code': status_code}  # e.g., NoSuchBucket = 404
+
+    def create_presigned_url(self, bucket_name, key_name, expiration_time=120):
+        """Creates a presigned URL to a BDBag stored in an S3 location.
+        
+        :param bucket_name: name of bucket in AWS account
+        :type bucket_name: str
+        :param key_name: object name in bucket on S3
+        :type key_name: str
+        :param expiration_time: expiration time of URL in seconds'
+        :type expiration_time: int
+        :return result: status code and presigned URL (in case of success)
+        :rtype: dict
+        """
+        result = {
+            'status_code': ''
+            }
+        status_code = self.bucket_exists(bucket_name)
+        if status_code == 200:
+            try:
+                url = self.bucket.generate_presigned_url(
+                    ClientMethod='get_object',
+                    Params={
+                        'Bucket': bucket_name,
+                        'Key': key_name
+                    },
+                    ExpiresIn=expiration_time
+                )
+            except ClientError as e:
+                result['status_code'] = \
+                    e.response['ResponseMetadata']['HTTPStatusCode']
+                result['error_code'] = e.response['Error']['Code']
+                result['msg'] = e.response['Error']['Message']
```

Comment:
Will `Error` always exist? I'm asking; I don't know.

Reply:
Looks like `Error` is always a key in the `ClientError` dictionary. See https://github.com/boto/botocore/blob/develop/botocore/exceptions.py#L377